Add BoundedMutuallyExclusiveWaitQueue

ian-axelrod opened the pull request:
BoundedMutuallyExclusiveWaitQueue--a clearly awesome name--adds to
go-waitqueue the ability to run multiple tasks concurrently up until a
threshold size, at which point the queue will block instead of
immediately return. BoundedMutuallyExclusiveWaitQueue also adds the idea
of mutually exclusive queues, which allows go-waitqueue to enforce
mutual exclusivity on tasks by their assigned name while respecting the
global size limit for the queue.

ian-axelrod committed:
Add BoundedMutuallyExclusiveWaitQueue

BoundedMutuallyExclusiveWaitQueue--a clearly awesome name--adds to
go-waitqueue the ability to run multiple tasks concurrently up until a
threshold size, at which point the queue will block instead of
immediately return. BoundedMutuallyExclusiveWaitQueue also adds the idea
of mutually exclusive queues, which allows go-waitqueue to enforce
mutual exclusivity on tasks by their assigned name while respecting the
global size limit for the queue.

mtabini commented on queue.go:
```diff
@@ -103,3 +103,44 @@ func (w *WaitQueue) Unlock() {
+}
+
+// New returns a new BoundedMutuallyExclusiveWaitQueue ready for use.
+func NewBounded(size int, queueName string, additionalQueueNames ...string) *BoundedMutuallyExclusiveWaitQueue {
```
There's no need to define both `queueName` and `additionalQueueNames` here. You can simply define a variadic parameter and check that the length of the resulting array > 0.

ian-axelrod replied:
Is golang's compiler smart enough to flag zero length variadic args + len check as a compile error if there are any calls to `NewBounded` that do not specify any queue names? I was under the assumption that it wasn't, so I wanted to ensure that we'd get a compile error in the aforementioned case rather than a runtime error.

mtabini replied:
It isn't. If that's what you're worried about, then your syntax is correct.

ian-axelrod replied:
Eh, not too worried about it. If you want me to change it I'll go ahead. If you're fine with it as is and think this is in a mergeable state, then go for the merge.

mtabini commented:
Aside from the one comment, this LGTM, but it would be great if you could update the comments so that they explain how the `BoundedMutuallyExclusiveWaitQueue` works. If you feel extra nice, you can also run `golint` against the code to make sure the comments follow semantic rules :-)

ian-axelrod commented:
What you're asking for is impossible. Impossible, I say! I'm on it :).

ian-axelrod committed:
Correct formatting + add comments

Corrected formatting in accordance with golint suggestions.

Added comments to all exported functions, and elaborated on existing
documentation.

ian-axelrod commented:
@mtabini Added more comments + golint corrections. Still waiting on your reply to https://github.com/mtabini/go-waitqueue/pull/1#discussion_r77847075 before making that change.
